src/soter/openssl/soter_engine.h

```diff
@@ -50,7 +50,8 @@ struct soter_rsa_key_pair_gen_type {
 };
 
 struct soter_asym_ka_type {
-    EVP_PKEY_CTX* pkey_ctx;
+    EVP_PKEY* param;
+    EVP_PKEY* pkey;
```

Comment:
this is some kind of dark magic - why on Earth key parameters are separate from the key itself (which is a meaningless string of bytes without parameters)?

Reply:
> why on Earth key parameters are separate from the key itself (which is a meaningless string of bytes without parameters)?

API design is hard. As it is, `soter_asym_ka_t` technically allows to first prepare for key generation with one curve and then import a key on some other curve, and who knows what's the expected curve should be when you generate after importing. `param` are only used to generate new keys, while `pkey` is either generated or imported key (with its own parameters embedded in it, the way we do imports). The `pkey` is what gets used for derivation.